## YES-NO Brain Tumor Detection Model

### First Iteration

The dataset is organised as depicted below. <br>


```
train_data/
├── NO
│   ├── image1.jpg
│   ├── image2.jpg
│   └── ...
└── YES
    ├── image1.jpg
    ├── image2.jpg
    └── ...
```



In [1]:
import os
import numpy as np
import tensorflow as tf
from tensorflow.keras import layers, models m n
from tensorflow.keras.preprocessing.image import ImageDataGenerator
from sklearn.model_selection import train_test_split

# Set your dataset paths
dataset_path = '/content/drive/MyDrive/CV/Brain Tumor Dataset/brain_tumor_dataset'

# Get the list of image files
yes_tumor_files = [os.path.join(dataset_path, 'YES', filename) for filename in os.listdir(os.path.join(dataset_path, 'YES'))]
no_tumor_files = [os.path.join(dataset_path, 'NO', filename) for filename in os.listdir(os.path.join(dataset_path, 'NO'))]

# Assuming all images have the same dimensions
img_height, img_width = 224, 224

# Create a list to store images and labels
images = []
labels = []

# Load images with tumors ('YES')
for file_path in yes_tumor_files:
    img = tf.keras.preprocessing.image.load_img(file_path, target_size=(img_height, img_width))
    img_array = tf.keras.preprocessing.image.img_to_array(img)
    images.append(img_array)
    labels.append(1)  # 1 for 'YES'

# Load images without tumors ('NO')
for file_path in no_tumor_files:
    img = tf.keras.preprocessing.image.load_img(file_path, target_size=(img_height, img_width))
    img_array = tf.keras.preprocessing.image.img_to_array(img)
    images.append(img_array)
    labels.append(0)  # 0 for 'NO'

# Convert the lists to NumPy arrays
images = np.array(images)
labels = np.array(labels)

# Split the data into training and validation sets
X_train, X_val, y_train, y_val = train_test_split(images, labels, test_size=0.2, random_state=42)

# Create a CNN model with more layers
model = models.Sequential()
model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(img_height, img_width, 3)))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(128, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Flatten())
model.add(layers.Dense(128, activation='relu'))
model.add(layers.Dropout(0.5))  # Dropout layer for regularization
model.add(layers.Dense(64, activation='relu'))
model.add(layers.Dense(1, activation='sigmoid'))

# Compile the model
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

# Data augmentation
train_datagen = ImageDataGenerator(rescale=1./255, shear_range=0.2, zoom_range=0.2, horizontal_flip=True)
train_datagen.fit(X_train)

# Train the model
model.fit(train_datagen.flow(X_train, y_train, batch_size=32), epochs=176, validation_data=(X_val, y_val))

# Evaluate the model
test_loss, test_acc = model.evaluate(X_val, y_val)
print(f"Test Accuracy: {test_acc}")

# Save the model
model.save("Brain Tumor Detection Model.h5")


SyntaxError: invalid syntax (ipython-input-4198241676.py, line 4)

### Trying to improve accuracy

Trying to improve the accuracy of the model by introducing some Data Augmentation to the images. <br>
The idea is Augmenting the images (altering them slighly) can improve the accuracy of the model. <br>
But, if not done correctly, it can actually lower the accuracy of the module. <br>


In [ ]:
import os
import numpy as np
import tensorflow as tf
from tensorflow.keras import layers, models
from tensorflow.keras.preprocessing.image import ImageDataGenerator
from sklearn.model_selection import train_test_split

# Set your dataset paths
dataset_path = '/content/drive/MyDrive/CV/Brain Tumor Dataset/brain_tumor_dataset'

# Get the list of image files
yes_tumor_files = [os.path.join(dataset_path, 'YES', filename) for filename in os.listdir(os.path.join(dataset_path, 'YES'))]
no_tumor_files = [os.path.join(dataset_path, 'NO', filename) for filename in os.listdir(os.path.join(dataset_path, 'NO'))]

# Assuming all images have the same dimensions
img_height, img_width = 224, 224

# Create a list to store images and labels
images = []
labels = []

# Load images with tumors ('YES')
for file_path in yes_tumor_files:
    img = tf.keras.preprocessing.image.load_img(file_path, target_size=(img_height, img_width))
    img_array = tf.keras.preprocessing.image.img_to_array(img)
    images.append(img_array)
    labels.append(1)  # 1 for 'YES'

# Load images without tumors ('NO')
for file_path in no_tumor_files:
    img = tf.keras.preprocessing.image.load_img(file_path, target_size=(img_height, img_width))
    img_array = tf.keras.preprocessing.image.img_to_array(img)
    images.append(img_array)
    labels.append(0)  # 0 for 'NO'

# Convert the lists to NumPy arrays
images = np.array(images)
labels = np.array(labels)

# Split the data into training and validation sets
X_train, X_val, y_train, y_val = train_test_split(images, labels, test_size=0.2, random_state=42)

# Create a CNN model with more layers
model = models.Sequential()
model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(img_height, img_width, 3)))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(128, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Flatten())
model.add(layers.Dense(128, activation='relu'))
model.add(layers.Dropout(0.5))  # Dropout layer for regularization
model.add(layers.Dense(64, activation='relu'))
model.add(layers.Dense(1, activation='sigmoid'))

# Compile the model
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

# Data augmentation
train_datagen = ImageDataGenerator(
    rescale=1./255,
    shear_range=0.2,
    zoom_range=0.2,
    horizontal_flip=True
)

# You can use the flow method directly on your image arrays
train_generator = train_datagen.flow(X_train, y_train, batch_size=32)

# Train the model
model.fit(train_generator, epochs=176, validation_data=(X_val, y_val))

# Evaluate the model
test_loss, test_acc = model.evaluate(X_val, y_val)
print(f"Test Accuracy: {test_acc}")

# Save the model
model.save("MRI Brain Tumor Detection Model v2.h5")

FileNotFoundError: [Errno 2] No such file or directory: '/content/drive/MyDrive/CV/Brain Tumor Dataset/brain_tumor_dataset/YES'

### Final Result

The accuracy of the original model was - 70.588% <br>
The accuracy of the  model after introducing data augmentation - 72.549% <br>
Accuracy was improved by 1.961%. <br>

## Advanced model cuz why  not (detection of multiply types of tumors)

This is just to test how we can go. This uses a Dataset that has MRI scans of 3 types of brain tumors and a set of normal brain scans. <br>
The types of brain tumors are Glioma, Meningioma, and Pituitary. They are labeled respectively in each set. <br>
The set of normal brain MRI scans is labeled 'notumor'

### Training a more Advanced Model

The dataset is organised in the manner depicted below, which is why the keyword 'categorical is used' <br>

```
train_data/
├── glioma
│   ├── image1.jpg
│   ├── image2.jpg
│   └── ...
├── meningioma
│   ├── image1.jpg
│   ├── image2.jpg
│   └── ...
├── notumor
│   ├── image1.jpg
│   ├── image2.jpg
│   └── ...
└── pituitary
    ├── image1.jpg
    ├── image2.jpg
    └── ...
```

In [ ]:
import tensorflow as tf
from tensorflow.keras.preprocessing.image import ImageDataGenerator
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Conv2D, MaxPooling2D, Flatten

# Load the MRI scan dataset
train_datagen = ImageDataGenerator(rescale=1./255)

train_generator = train_datagen.flow_from_directory(
    '/content/drive/MyDrive/CV/Brain Tumor Dataset/advanced_brain_tumor_dataset/Training',  # Directory containing training MRI scans
    target_size=(224, 224),  # Resize all images to 224x224 pixels
    batch_size=32,  # Number of images per batch
    class_mode='categorical'  # Classify images into four categories (glioma, meningioma, notumor, pituitary)
)

# Define the CNN model architecture
model = Sequential()

model.add(Conv2D(32, (3, 3), activation='relu', input_shape=(224, 224, 3)))
model.add(MaxPooling2D((2, 2)))

model.add(Conv2D(64, (3, 3), activation='relu'))
model.add(MaxPooling2D((2, 2)))

model.add(Conv2D(128, (3, 3), activation='relu'))
model.add(MaxPooling2D((2, 2)))

model.add(Flatten())
model.add(Dense(256, activation='relu'))
model.add(Dense(512, activation='relu'))
model.add(Dense(4, activation='softmax'))  # Four classes (glioma, meningioma, notumor, pituitary)


# Compile the model
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

# Train the model
model.fit(train_generator, epochs=10)

# Save the model
model.save("Advanced Brain Tumor Detection Model.h5")

Found 5712 images belonging to 4 classes.
Epoch 1/10
179/179 [==============================] - 640s 4s/step - loss: 0.5808 - accuracy: 0.7589
Epoch 2/10
179/179 [==============================] - 597s 3s/step - loss: 0.2214 - accuracy: 0.9228
Epoch 3/10
179/179 [==============================] - 603s 3s/step - loss: 0.1348 - accuracy: 0.9519
Epoch 4/10
179/179 [==============================] - 622s 3s/step - loss: 0.0617 - accuracy: 0.9799
Epoch 5/10
179/179 [==============================] - 600s 3s/step - loss: 0.0412 - accuracy: 0.9870
Epoch 6/10
179/179 [==============================] - 601s 3s/step - loss: 0.0432 - accuracy: 0.9867
Epoch 7/10
179/179 [==============================] - 601s 3s/step - loss: 0.0263 - accuracy: 0.9909
Epoch 8/10
179/179 [==============================] - 598s 3s/step - loss: 0.0120 - accuracy: 0.9951
Epoch 9/10
179/179 [==============================] - 598s 3s/step - loss: 0.0166 - accuracy: 0.9951
Epoch 10/10
179/179 [============================

/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3079: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


### Testing the Model


In [ ]:
test_datagen = ImageDataGenerator(rescale=1./255)

test_generator = test_datagen.flow_from_directory(
    '/content/drive/MyDrive/CV/Brain Tumor Dataset/advanced_brain_tumor_dataset/Testing',  # Directory containing testing MRI scans
    target_size=(224, 224),  # Resize all images to 224x224 pixels
    batch_size=32,  # Number of images per batch
    class_mode='categorical'  # Classify images into four categories (glioma, meningioma, notumor, pituitary)
)

scores = model.evaluate(test_generator)
print("Test loss:", scores[0])
print("Test accuracy:", scores[1])


Found 1311 images belonging to 4 classes.
41/41 [==============================] - 340s 8s/step - loss: 0.2285 - accuracy: 0.9542
Test loss: 0.2285175621509552
Test accuracy: 0.9542334079742432


### Final Result

The model recieved a Test Accuracy of 95.423%. <br>
The difference in accuracy between this model and the previous model could be due to the larger dataset, giving the second model more room to learn and grow.

# Trained Models and Dataset

The Dataset, as well as the model have been provided in the following Google Drive Link <br>
GDrive - https://drive.google.com/drive/folders/1-0DtpFBkZwcdMXwOcX1pORdqCpH5Nn63?usp=drive_link